sdk/identity/azure-identity/src/azure_cli_credential.cpp

```diff
@@ -0,0 +1,654 @@
+#include <utility>
+#include <vector>
+
+#if defined(AZ_PLATFORM_WINDOWS)
+#if !defined(WIN32_LEAN_AND_MEAN)
+#define WIN32_LEAN_AND_MEAN
+#endif
+#if !defined(NOMINMAX)
+#define NOMINMAX
+#endif
+#include <windows.h>
+#else
+#include <errno.h>
+#include <fcntl.h>
+#include <signal.h>
+#include <spawn.h>
+#include <sys/wait.h>
+#include <unistd.h>
+#endif
+
+using Azure::Identity::AzureCliCredential;
+
+using Azure::DateTime;
+using Azure::Core::Context;
+using Azure::Core::_internal::Environment;
+using Azure::Core::Credentials::AccessToken;
+using Azure::Core::Credentials::AuthenticationException;
+using Azure::Core::Credentials::TokenCredentialOptions;
+using Azure::Core::Credentials::TokenRequestContext;
+using Azure::Identity::AzureCliCredentialOptions;
+using Azure::Identity::_detail::TokenCredentialImpl;
+
+namespace {
+void ThrowIfNotSafeCmdLineInput(std::string const& input, std::string const& description)
+{
+  for (auto const c : input)
+  {
+    switch (c)
+    {
+      case ':':
+      case '/':
+      case '.':
+      case '-':
+      case '_':
+      case ' ':
+        break;
+
+      default:
+        if (!std::isalnum(c))
+        {
+          throw AuthenticationException(
+              "AzureCliCredential: Unsafe command line input found in " + description + ": "
+              + input);
+        }
+    }
+  }
+}
+} // namespace
+
+AzureCliCredential::AzureCliCredential(
+    std::string tenantId,
+    DateTime::duration cliProcessTimeout,
+    Core::Credentials::TokenCredentialOptions const& options)
+    : m_tenantId(std::move(tenantId)), m_cliProcessTimeout(std::move(cliProcessTimeout))
+{
+  static_cast<void>(options);
+  ThrowIfNotSafeCmdLineInput(m_tenantId, "TenantID");
+}
+
+AzureCliCredential::AzureCliCredential(AzureCliCredentialOptions const& options)
+    : AzureCliCredential(options.TenantId, options.CliProcessTimeout, options)
+{
+}
+
+AzureCliCredential::AzureCliCredential(TokenCredentialOptions const& options)
+    : AzureCliCredential(
+        AzureCliCredentialOptions{}.TenantId,
+        AzureCliCredentialOptions{}.CliProcessTimeout,
+        options)
+{
+}
+
+std::string AzureCliCredential::GetAzCommand(
+    std::string const& resource,
+    std::string const& tenantId) const
+{
+  ThrowIfNotSafeCmdLineInput(resource, "Resource");
+  std::string command = "az account get-access-token --output json --resource \"" + resource + "\"";
+
+  if (!tenantId.empty())
+  {
+    command += " --tenant \"" + tenantId + "\"";
+  }
+
+  return command;
+}
+
+namespace {
+std::string RunShellCommand(
+    std::string const& command,
+    DateTime::duration timeout,
+    Context const& context);
+}
+
+AccessToken AzureCliCredential::GetToken(
+    TokenRequestContext const& tokenRequestContext,
+    Context const& context) const
+{
+  try
+  {
+    auto const azCliResult = RunShellCommand(
+        GetAzCommand(
+            TokenCredentialImpl::FormatScopes(tokenRequestContext.Scopes, true, false), m_tenantId),
```

Comment:
consider using `--scopes` instead of `resource`

Reply:
FWIF, other SDKs do use the resource, but I like the scopes, thanks for the info! I'll try to update it. The credential will actually be more capable with scopes, so I like it even more.

Reply:
.NET and Java SDKs are consistent about having it as a resource.
I am going to leave it using resource in this PR, but I'll create a work item to follow up, and implement `--scope` unless there is a strong reason why other SDKs are not doing it. I also will need to test it, scopes come in many formats, and I want to make sure we support them well.

Reply:
Thanks for the note @vhvb1989, I put this into issue for further exploration: #4153
I'd love to add scope support, as it is a superset of current functionality, but I need to collect more data about it (I put more details in the issue that I linked). Not in this PR.